File: src/lib/session.ts
import "server-only";
import { createHmac, timingSafeEqual } from "crypto";
import { cookies } from "next/headers";

const COOKIE_NAME = "yanaloma_session";

// Se lee en tiempo de ejecución (no al cargar el módulo) para que el build
// de Next.js no falle recolectando datos de página antes de que exista el
// secret en el entorno (recién se inyecta como Fly secret en runtime).
function getSecret(): string {
  const secret = process.env.SESSION_SECRET;
  if (!secret) {
    throw new Error("Falta SESSION_SECRET en las variables de entorno.");
  }
  return secret;
}

export type SessionUser = {
  id: number;
  username: string;
  nombre: string;
  role: "admin" | "cajero" | string;
};

function sign(payload: string): string {
  return createHmac("sha256", getSecret()).update(payload).digest("hex");
}

function encode(user: SessionUser): string {
  const payload = Buffer.from(JSON.stringify(user)).toString("base64url");
  const signature = sign(payload);
  return `${payload}.${signature}`;
}

function decode(value: string): SessionUser | null {
  const [payload, signature] = value.split(".");
  if (!payload || !signature) return null;

  const expected = sign(payload);
  const a = Buffer.from(signature);
  const b = Buffer.from(expected);
  if (a.length !== b.length || !timingSafeEqual(a, b)) return null;

  try {
    return JSON.parse(Buffer.from(payload, "base64url").toString("utf8"));
  } catch {
    return null;
  }
}

export async function setSession(user: SessionUser) {
  const store = await cookies();
  store.set(COOKIE_NAME, encode(user), {
    httpOnly: true,
    secure: process.env.NODE_ENV === "production",
    sameSite: "lax",
    path: "/",
    maxAge: 60 * 60 * 12, // 12 horas, un turno largo
  });
}

export async function getSession(): Promise<SessionUser | null> {
  const store = await cookies();
  const value = store.get(COOKIE_NAME)?.value;
  if (!value) return null;
  return decode(value);
}

export async function clearSession() {
  const store = await cookies();
  store.delete(COOKIE_NAME);
}

File: app/session.py
import base64
import hashlib
import hmac
import json
import os

from fastapi import Response

COOKIE_NAME = "yanaloma_session"


def _secret() -> bytes:
    secret = os.environ.get("SESSION_SECRET")
    if not secret:
        raise RuntimeError("Falta SESSION_SECRET en las variables de entorno.")
    return secret.encode()


def _b64url_encode(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).rstrip(b"=").decode()


def _b64url_decode(value: str) -> bytes:
    padding = "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(value + padding)


def _sign(payload: str) -> str:
    return hmac.new(_secret(), payload.encode(), hashlib.sha256).hexdigest()


def encode_session(user: dict) -> str:
    payload = _b64url_encode(json.dumps(user).encode())
    return f"{payload}.{_sign(payload)}"


def decode_session(value: str) -> dict | None:
    try:
        payload, signature = value.split(".", 1)
    except ValueError:
        return None

    if not hmac.compare_digest(signature, _sign(payload)):
        return None

    try:
        return json.loads(_b64url_decode(payload))
    except Exception:
        return None


def set_session_cookie(response: Response, user: dict) -> None:
    response.set_cookie(
        COOKIE_NAME,
        encode_session(user),
        httponly=True,
        samesite="lax",
        secure=os.environ.get("ENV") == "production",
        path="/",
    )


def clear_session_cookie(response: Response) -> None:
    response.delete_cookie(COOKIE_NAME, path="/")
